### PR TITLE
Clarify behavior of RayCast when `get_collision_point()` is used inside a collision shape

### PR DESCRIPTION
**Pull request by CardboardCarl**
Clarified documentation on a behavior of get_collision_point() in both RayCast2D and RayCast3D objects when it is used while the ray is inside a collision shape

### CODE REVIEW

**Comment by MewPurPur** [comment]
Here it should maybe just say "See also [member hit_from_inside]" as it tells explains the same thing verbatim, and also it explains the limitation with concave shapes.

File: doc/classes/RayCast2D.xml [hunk 1]
```diff
@@ -74,7 +74,7 @@
 		<method name="get_collision_point" qualifiers="const">
 			<return type="Vector2" />
 			<description>
-				Returns the collision point at which the ray intersects the closest object.
+				Returns the collision point at which the ray intersects the closest object. If [code]hit_from_inside[/code] is [code]true[/code] and the ray starts inside of a collision shape, this function will return the origin point of the ray.
```
Comment (AThousandShips) [inline doc/classes/RayCast2D.xml:77]:
```suggestion
				Returns the collision point at which the ray intersects the closest object. If [member hit_from_inside] is [code]true[/code] and the ray starts inside of a collision shape, this function will return the origin point of the ray.
```

File: doc/classes/RayCast3D.xml [hunk 2]
```diff
@@ -81,7 +81,7 @@
 		<method name="get_collision_point" qualifiers="const">
 			<return type="Vector3" />
 			<description>
-				Returns the collision point at which the ray intersects the closest object.
+				Returns the collision point at which the ray intersects the closest object. If [code]hit_from_inside[/code] is [code]true[/code] and the ray starts inside of a collision shape, this function will return the origin point of the ray.
```
Comment (AThousandShips) [inline doc/classes/RayCast3D.xml:84]:
```suggestion
				Returns the collision point at which the ray intersects the closest object. If [member hit_from_inside] is [code]true[/code] and the ray starts inside of a collision shape, this function will return the origin point of the ray.
```

**Comment by CardboardCarl** [comment]
> Here it should maybe just say "See also [member hit_from_inside]" as it explains the same thing verbatim, and also it explains the limitation with concave shapes.

The issue with this is that the returned value is not Vector2(0, 0), but instead it's the global position of the ray, so the behavior is slightly different.

**Comment by CardboardCarl** [comment]
Oops! Going to rebase this in a little bit.

**Comment by CardboardCarl** [comment]
Changes have been rebased into one commit. :+1:

**Comment by AThousandShips** [comment]
Someone knowledgeable on the area will come and review this, was just fixing errors in the docs that caused issues

**Comment by CardboardCarl** [comment]
> Someone knowledgeable on the area will come and review this, was just fixing errors in the docs that caused issues

No worries, thanks for checking in!

File: doc/classes/RayCast2D.xml [hunk 3]
```diff
@@ -74,7 +74,7 @@
 		<method name="get_collision_point" qualifiers="const">
 			<return type="Vector2" />
 			<description>
-				Returns the collision point at which the ray intersects the closest object.
+				Returns the collision point at which the ray intersects the closest object. If [member hit_from_inside] is [code]true[/code] and the ray starts inside of a collision shape, this function will return the origin point of the ray.
```
Comment (Calinou) [inline doc/classes/RayCast2D.xml:77]: Are we certain that the *exact* return value is the origin point of the ray? People may be trying to perform exact comparison to check if the ray was inside something, and we need to make sure this assertion is correct.

If this is not the case, I suggest rewording it to mention that the position will be similar, but not necessarily identical to the ray origin.

Comment (CardboardCarl) [reply doc/classes/RayCast2D.xml:77]: Every time I tried it, it was always the exact same point, even with floating point positions. I'll try to make a test scene in a little bit to do some further testing.

Comment (CardboardCarl) [reply doc/classes/RayCast2D.xml:77]: Made a script to test this out. It's always EXACTLY the same point. You can test for yourself if you'd like. Just add this to an empty Node2D in an empty scene:

```
extends Node2D

var sample_size: int = 20
@onready var static_body: StaticBody2D = StaticBody2D.new()
@onready var collision_shape: CollisionShape2D = CollisionShape2D.new()
@onready var raycast: RayCast2D = RayCast2D.new()

func _ready():
	collision_shape.shape = RectangleShape2D.new()
	collision_shape.shape.size = Vector2(PI * 10, PI * 10)
	raycast.enabled = true
	raycast.hit_from_inside = true
	raycast.collide_with_bodies = true
	raycast.collide_with_bodies = true
	add_child(static_body)
	add_child(raycast)
	static_body.call_deferred("add_child", collision_shape)
	static_body.call_deferred("set", "global_position", Vector2(collision_shape.shape.size.x / 2, collision_shape.shape.size.y / 2))
	var timer = get_tree().create_timer(1)
	timer.connect("timeout", Callable(self, "call_timer"))

func start_test():
	for i in range(0, sample_size):
		randomize()
		raycast.position = Vector2(randf_range(0, collision_shape.shape.size.x), randf_range(0, collision_shape.shape.size.y))
		randomize()
		raycast.target_position = Vector2(randf_range(-100, 100), randf_range(-100, 100))
		raycast.force_raycast_update()
		if raycast.is_colliding():
			var collision_point = raycast.get_collision_point()
			print("=================================================")
			print("ATTEMPT %s: \n RAY POSITION: %s (RAY TARGET: %s) \n COLLISION POINT: %s \n TEST_SUCCESS: %s" % [i, raycast.global_position, raycast.target_position, collision_point, bool(collision_point == raycast.global_position)])
			print("=================================================")
		else:
			print("(Attempt %s) ERROR: Raycast is not colliding!" % i)

func call_timer():
	start_test()

```

**Review by Mickeon — APPROVED** [review]
Very simple but nice addition. Should probably be cherrypicked for 4.2, too.

**Review by akien-mga — APPROVED** [review]
Language looks good. I'm assuming the doc change is actually correct, didn't double check :)

**Comment by YuriSizov** [comment]
Could you please squash your commits into one? Make sure that the final commit has a short but descriptive message (the title of this PR is a good option). See [this documentation](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase), if you need help with squashing.

**Comment by CardboardCarl** [comment]
@YuriSizov very sorry about that, I could have sworn I did earlier! Working on it.

**Comment by CardboardCarl** [comment]
Had to reorganize a little bit. @YuriSizov commits have been squashed!

**Review by YuriSizov — APPROVED** [review]

**Comment by YuriSizov** [comment]
Thanks! And congrats on your first merged Godot contribution!

**Comment by YuriSizov** [comment]
Cherry-picked for 4.2.2.
